src/trace_ng/log.zig

```diff
@@ -0,0 +1,672 @@
+
+        pub fn deinit(self: *Self) void {
+            self.recycle_fba.deinit();
+            if (self.log_msg) |log_msg| {
+                log_msg.deinit();
+            }
+        }
+
+        pub fn log(self: *Self, level: Level, message: []const u8) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                message,
+                null,
+                null,
+                null,
+            );
+            const writer = self.log_msg.?.writer();
+            logfmt.writeLog(writer, log_message) catch @panic("Failed to write log");
+        }
+
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                message,
+                fields,
+                null,
+                null,
+            );
+            const writer = self.log_msg.?.writer();
+            logfmt.writeLog(writer, log_message) catch @panic("Failed to write log");
+        }
+
+        pub fn logf(self: *Self, level: Level, comptime fmt: []const u8, args: anytype) void {
+            if (@intFromEnum(level) > @intFromEnum(self.max_level)) {
+                // noop
+                return;
+            }
+            const maybe_scope = blk: {
+                if (scope) |s| {
+                    break :blk @typeName(s);
+                } else {
+                    break :blk null;
+                }
+            };
+
+            const log_message = logfmt.createLogMessage(
+                &self.recycle_fba,
+                self.max_buffer,
+                level,
+                maybe_scope,
+                null,
+                null,
+                fmt,
+                args,
+            );
+            const writer = self.log_msg.?.writer();
+            logfmt.writeLog(writer, log_message) catch @panic("Failed to write log");
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
```

Comment:
There is something weird about having two anytypes back to back. I'm not sure what is best. What would you think about putting `fields` before `fmt`? This way the code that calls this method would actually look more similar to how usages of `.field()` currently looks.

Reply:
So revising this now, and for example we currently have

```
   logger.errfWithFields(
        "{s}",
        .{"Logging with logfWithFields"},
        .{
            .f_agent = "Firefox",
            .f_version = 120,
            .f_local = "en",
            .f_stock = "nvidia",
        },
    );
```

I can move the fields to to be the first argument,  which then becomes:

```
   logger.errfWithFields(.{
            .f_agent = "Firefox",
            .f_version = 120,
            .f_local = "en",
            .f_stock = "nvidia",
        },
        "{s}",
        .{"Logging with logfWithFields"},
    );
```

This is without renaming. For renaming it would then be something like

```
   logger.fieldsErrf(.{
            .f_agent = "Firefox",
            .f_version = 120,
            .f_local = "en",
            .f_stock = "nvidia",
        },
        "{s}",
        .{"Logging with logfWithFields"},
    );
```

only that this won't be consistent with the other naming `err`, `errf` where the `err` (or `info`, or `debug` etc) comes first.

What do you think? We move the fields to the first but keep the naming?